lib/charms/dns_transfer/v0/dns_transfer.py

```diff
@@ -0,0 +1,346 @@
+# Copyright 2025 Canonical Ltd.
+# Licensed under the Apache2.0. See LICENSE file in charm source for details.
+# pylint: disable=R0801
+"""Library to manage the integration between primary and secondary DNS charms.
+
+This library contains the Requires and Provides classes for handling the integration
+between an application and a charm providing the `dns_record` integration.
+
+### Requirer Charm
+
+The DNSTransferRequires provides:
+
+* an `get_remote_relation_data" method to get data from the provider.
+* an `update_relation_data` method to update the relation data by
+passing a list of IP addresses to the provider.
+
+```python
+
+from charms.dns_transfer.v0.dns_transfer import DNSTransferRequires, DNSTransferRequirerData
+
+class DNSTransferRequirerCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.dns_transfer = DNSTransferRequires(self)
+        # line-too-long
+        self.framework.observe(charm.on["dns-transfer"].relation_changed, self._on_relation_changed) # noqa: W505 pylint: disable=C0301
+        ...
+
+    def _on_relation_changed(self, event: ops.Event) -> None:
+        dns_primary_data = self.dns_transfer.get_remote_relation_data()
+        addresses = dns_primary_data.addresses
+
+    def _on_config_changed(self, event: ops.Event) -> None:
+        addresses = self.my_ips()
+        relation = self.model.get_relation(self.dns_transfer.relation_name)
+        requirer_data = DNSTransferRequirerData(addresses=addresses)
+        self.dns_transfer.update_relation_data(relation, requirer_data)
+
+```
+
+### Provider Charm
+
+Following the previous example, this is an example of the provider charm.
+
+```python
+from charms.dns_transfer.v0.dns_transfer import DNSTransferProvides, DNSTransferProviderData
+
+class DNSTransferProviderCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.dns_transfer = DNSTransferProvides(self)
+        self.framework.observe(charm.on["dns-transfer"].relation_changed, self._on_relation_changed)
+        ...
+
+    def _on_relation_changed(self, event: ops.Event) -> None:
+        addresses = []
+        for relation in self.model.relations[self.dns_transfer.relation_name]:
+            dns_secondary_data = self.dns_transfer.get_remote_relation_data(relation)
+            addresses.extend(dns_secondary_data.addresses)
+
+    def _on_config_changed(self, event: ops.Event) -> None:
+        provider_data = DNSTransferProviderData()
```

Comment:
Wouldn't it be better to use `DNSTransferProviderData.model_validate()` to instantiate it ?